File: runtime/ruby/units/wasi/fd_filestat_set_size.rb
# requires: wasi/rights
def wasi_fd_filestat_set_size(fd, size)
  io = @fds[fd]
  return ERRNO_BADF if io.nil? || io.is_a?(WasiDir)
  return ERRNO_NOTCAPABLE unless fd_has_right?(fd, RIGHT_FD_FILESTAT_SET_SIZE)
  io.truncate(size)
  ERRNO_SUCCESS
rescue SystemCallError
  ERRNO_IO
end

File: runtime/perl/units/wasi/fd_filestat_set_size.pl
# requires:
sub wasi_fd_filestat_set_size {
    my ($self, $fd, $size) = @_;
    my $e = $self->{fds}{$fd};
    return ERRNO_BADF if !defined($e) || $e->{dir};
    return ERRNO_NOTCAPABLE unless $self->{meta}{$fd}[0] & RIGHTS_FD_FILESTAT_SET_SIZE;
    return ERRNO_IO unless truncate($e->{fh}, $size);
    return ERRNO_SUCCESS;
}
